common/src/main/kotlin/org/quizfight/common/Connection.kt

```diff
@@ -1,12 +1,44 @@
 package org.quizfight.common
 
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
 import org.quizfight.common.messages.Message
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+import java.net.ServerSocket
+import java.net.Socket
 import kotlin.reflect.KClass
 
-class Connection(private val handler: Map<KClass<*>, (Message) -> Unit>) {
+interface Connection {
+    fun send(msg: Message)
+    val handlers: Map<KClass<*>, (Connection, Message) -> Unit>
+}
 
+class SocketConnection(
+    val socket: Socket,
+    override val handlers: Map<KClass<*>, (Connection, Message) -> Unit>
+) : Connection {
+    // Do not switch order of stream creation, will result in deadlock
+    private val outStream = ObjectOutputStream(socket.getOutputStream())
+    private val inStream  = ObjectInputStream(socket.getInputStream())
 
-    fun send(msg: Message) {
+    init {
+        /* TODO: Type check handler map
+            for (key in handler.keys)
+            if (Message::class in key.supertypes)
+         */
+        receiveAsync()
+    }
 
+    private fun receiveAsync() = GlobalScope.launch {

```

Comment:
GlobalScope.launch could become an issue because its scope is the application's lifetime. Think of a game server running for weeks, a total of 100.000 players connecting. 2.000 of them stop with an exception / disconnect unexpectedly. The coroutine won't notice that the socket is no longer connected. That's 2.000 coroutines that are never terminated. I might be wrong though 👶 

https://stackoverflow.com/questions/54335365/why-not-use-globalscope-launch
https://medium.com/@elizarov/structured-concurrency-722d765aa952